### PR TITLE
[SPARK-22797][PySpark] Bucketizer support multi-column

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
Bucketizer support multi-column in the python side

## How was this patch tested?
existing tests and added tests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
ping @holdenk  , can you help reviewing this?

File: python/pyspark/ml/feature.py
Comment (viirya) [inline python/pyspark/ml/feature.py]: Update this doc too?

File: python/pyspark/ml/param/__init__.py
Comment (viirya) [inline python/pyspark/ml/param/__init__.py]: `toListFloat` requires each list entry is a numeric `TypeConverters._is_numeric`. Should `toListListFloat` have such requirement?

**Comment by viirya** [comment]
This needs an individual JIRA. @MLnick created SPARK-22797 for this. Please use it.

**Comment by zhengruifeng** [comment]
@viirya Thanks a lot for reviewing this! I will update the title to use the new ticket.

File: python/pyspark/ml/feature.py
Comment (viirya) [inline python/pyspark/ml/feature.py]: Note: there is a work to change this behavior to throw an exception, instead of a log warning. Remember to change this document later.

**Comment by viirya** [comment]
One minor comment, otherwise LGTM.

**Comment by zhengruifeng** [comment]
ping @MLnick ?

File: python/pyspark/ml/feature.py
Comment (MLnick) [inline python/pyspark/ml/feature.py]: Perhaps it would be cleaner to do a `df.show()` here? Likewise above for `bucketed` we could change that part of the doctest too.

File: python/pyspark/ml/feature.py
Comment (MLnick) [inline python/pyspark/ml/feature.py]: `values1` & `values2`?

File: python/pyspark/ml/param/__init__.py
Comment (MLnick) [inline python/pyspark/ml/param/__init__.py]: We need a test case in `ParamTypeConversionTests` for this new method; see `test_list_float` for reference.

**Review by MLnick — CHANGES_REQUESTED** [review]

**Comment by zhengruifeng** [comment]
@MLnick  Thanks for your reviewing and suggestions. I have updated this PR

File: python/pyspark/ml/feature.py
Comment (MLnick) [inline python/pyspark/ml/feature.py]: It may actually be neater to show only `values1` and `bucketed` - so perhaps `.transform(df.select('values1'))`?

**Review by MLnick — CHANGES_REQUESTED** [review]
One minor comment on the updated doctest.

I don't think this will make it into `2.3` given the code freeze and branch has been cut already. In which case we will need to change the `@since` tags.

Pending the error throwing PR for Scala `Bucketizer`, we can update the doc here.

**Comment by holdenk** [comment]
I mean I think it might have a chance, generally speaking we've allowed outstanding PRs to be merged after the freeze. Since there are outstanding blockers on the branch preventing us from cutting RC2 maybe its ok to move forward if we can do it quickly? Of course I defer to MLNick :)

**Comment by MLnick** [comment]
I’m generally ok with these small python api wrapper additions getting
merged as long as the risk of breaking anything is low - and here it is
since it’s just api parity

**Review by holdenk — APPROVED** [review]
LGTM and I'll merge today (Australia time) if there are no objections. (note: this means also waiting on @MLnick switching from -1 to approve).

**Comment by MLnick** [comment]
If it is going to get merged to `branch-2.3` the `since` tags need to be `2.3.0` again

**Comment by holdenk** [comment]
@MLnick you ok with this then?

File: python/pyspark/ml/feature.py
Comment (MLnick) [reply python/pyspark/ml/feature.py]: @holdenk this comment will need to be changed as per #19993 - but that has not been merged yet. I think #19993 will block 2.3 though, so we could preemptively change the doc here to match the Scala side in #19993 about throwing and exception.

**Comment by MLnick** [comment]
RC2 has been cut - @jkbradley do you see #19993 as a blocker? I think it should be merged for `2.3`. And also there are QA JIRAs (sub-tasks of [SPARK-23105](https://issues.apache.org/jira/browse/SPARK-23105)) that are blockers that are not reflected in the list of blockers for `2.3` as they are not targeted.

**Comment by MLnick** [comment]
Merged to master / branch-2.3. Thanks!

**Comment by MLnick** [comment]
I reverted this (see #20410 for details) - we can re-open it once that issue is solved.

**Comment by viirya** [comment]
Can this be re-open now?

**Comment by its-sherry** [comment]
Has there been any update on re-opening this? And also adding multiple column support to QuantileDiscretizer?

**Comment by zhengruifeng** [comment]
I am sorry, I am afread I can not re-open this PR beacuse I deleted it by mistake.
I just create another PR https://github.com/apache/spark/pull/25801 to continue.

**Pull request closed without merging**
